Dockerfile

```diff
@@ -0,0 +1,30 @@
+FROM ghcr.io/rse-ops/cuda-ubuntu-20.04:cuda-11.0.3
+
+# docker build -t ghcr.io/mfem/mfem-ubuntu .
+
+RUN apt-get update && \
+    apt-get install -y unzip gfortran && \ 
+    spack compiler find && \
+    apt-get install -y libcurl4-openssl-dev libssl-dev
+
+# /code is the working directory for code
+WORKDIR /code
+COPY . /code
+
+# This is for a spack environment/view to install from there
+WORKDIR /opt/mfem-env
+RUN . /opt/spack/share/spack/setup-env.sh && \
+    spack env create -d . && \
+    echo "  concretization: together" >> spack.yaml && \
+    spack env activate . && \
+    spack develop --path /code mfem@master+examples+miniapps && \
+    spack add mfem@master+examples+miniapps && \
+    spack install
+
+# ensure mfem always on various paths
+RUN cd /opt/mfem-env && \
+    spack env activate --sh -d . >> /etc/profile.d/z10_spack_environment.sh
```

Comment:
This (`z10_spack_environment.sh`) is a bit cryptic, can you elaborate?

Reply:
Yes! So what that does it write a file into /etc/profile.d, which will be sourced when you shell into the container. This ensures that when you shell into the container, all the mfem deps / binaries / libs are on the path and ready to go, and the user doesn't need to confusingly figure it out. If you don't add this, you have to instruct the reader to cd to the environment, run the activation command, etc.

Reply:
Is the name standard or hard-coded somewhere? Sorry, I haven't seen that before.

Reply:
The files are sourced based on their ABC name, so a z10 is very likely to be sourced last. The rest of the name doesn't really matter except to identify what the file is doing.

Reply:
It's something I learned developing for Singularity - if you looked in the ~/.singularity/env folder you'd find a bunch of named files with that convention. https://github.com/sylabs/singularity/tree/vault/release-2.4/libexec/bootstrap-scripts/environment. We would add (to that base) files to source for apps and docker, and the naming would determine the order.

Reply:
Got it.

I'm learning several new things today 👍